### PR TITLE
Create add_contributor.yml

### LINKED ISSUE #10885: create Github Action to automatically update AUTHORS.md
## goal
create a new github action to incorporate into our workflow which would add a new contributor's name to [AUTHORS.md](https://github.com/learningequality/kolibri/blob/develop/AUTHORS.md) as part of their first pull request. 

## why?
currently, this element of the PR/contribution process is very often missed, requiring contributors' names to be added manually & separately from their contribution, which has led to some contributors not getting added to the file at all.

## potential resources
there are existing github actions available on github marketplace like [auto-add-contributors](https://github.com/marketplace/actions/auto-add-contributors) which could provide a helpful reference point for what we want to accomplish, but for our action we also need to assess the current content of `AUTHORS.md` to avoid repeat mentions of the same contributor, and ideally would also scrape a contributor's full name from their gh profile, in order to record that in addition to their username.

### PR DESCRIPTION
**Pull request by GarvitSinghal47**
## Summary

This PR adds a GitHub Actions workflow to automatically add contributors to the AUTHORS.md file when a pull request is opened or synchronized. The workflow checks if the contributor already exists in AUTHORS.md, retrieves the full name and username of the new contributor, and appends a new table row in the format "| FULL_NAME | USERNAME |" to maintain the desired formatting. The changes in this PR align with the new GitHub standards for set-output and ensure proper formatting of the AUTHORS.md file.

…

## References

Fixes #10885

…

## Reviewer guidance

…

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: .github/workflows/add_contributor.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+
+      - name: Check if contributor already exists
+        id: check_contributor
+        run: |
+          exists=$(grep -c -F "$GITHUB_ACTOR" AUTHORS.md)
+          if [[ $exists -gt 0 ]]; then
+            echo "::set-output name=exists::true"
```
Comment (thanksameeelian) [inline .github/workflows/add_contributor.yml]: github has deprecated `set-output` for actions - please [see here](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/) for more info & examples of how to refactor code to align with their new standards. would you be able to make similar updates to your code here?

File: .github/workflows/add_contributor.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          if [[ -z "$full_name" ]]; then
+            full_name=$username
+          fi
+          echo "::set-output name=full_name::$full_name"
+
+      - name: Append contributor to AUTHORS.md
+        id: append_contributor
+        if: steps.check_contributor.outputs.exists == 'false'
+        run: |
+          echo "$GITHUB_ACTOR - ${{ steps.get_full_name.outputs.full_name }}" >> AUTHORS.md
```
Comment (thanksameeelian) [inline .github/workflows/add_contributor.yml]: our [AUTHORS.md](https://github.com/learningequality/kolibri/blob/develop/AUTHORS.md) file is in a specific format we'd like to maintain: a new table row for each entry with a first column of `FULL_NAME` & a second column of `USERNAME`, in that specific order. could you update the output here to reflect that order & include the `|`s necessary for proper formatting? 

for instance, [the code that produces the table row for my own entry](https://github.com/learningequality/kolibri/blob/a196c1ee581481670ca1400df1ccaf6bf8ba8053/AUTHORS.md?plain=1#L98) is:
`| Amelia Breault | thanksameeelian |`

**Review by thanksameeelian — COMMENTED** [review]
thank you so much for your contribution, @GarvitSinghal47!

would you be able to flesh out the description of the PR a bit? we don't need much, but it would be helpful to add a sentence or two to "Summary" to provide context for what this PR is trying to address. if you include `fixes #10885` in the "References" section, this PR can be more tightly linked to its related issue (which is helpful for reviewers but also can facilitate github automatically closing that issue once this gets merged).

i've left a couple comments requesting small changes, and once those are addressed i think this PR will be in great shape. please feel free to ping me in a comment or re-request my review in the sidebar once those updates are implemented! 🙂

**Comment by rtibbles** [comment]
Hrm any idea why the new action is failing? It seems to be happening at the step that tries to set the output.

**Comment by GarvitSinghal47** [comment]
> Hrm any idea why the new action is failing? It seems to be happening at the step that tries to set the output.

tried something after reading some docs related to assigning of variable issue or output setting , lets see if it works or not .

**Comment by thanksameeelian** [comment]
@GarvitSinghal47, it seems like there are a couple things happening here that are continuing to cause snags. 

1. it appears like throughout this proposed action, the output is not getting sent to the correct environment file. [the github post about deprecating `set-output`](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/) provides guidance for using 
```echo "{name}={value}" >> $GITHUB_OUTPUT```, with `$GITHUB_OUTPUT` as the name of the new environment file we should be accessing

2. it appears like current line 45, `echo "| $full_name | $GITHUB_ACTOR |" >> AUTHORS.md`, may not be working as intended because we don't have access to the value of `$full_name` in this context, as [that value is assigned in a separate `step`](https://docs.github.com/en/actions/learn-github-actions/contexts#steps-context). but the value can be accessed via the method explained in the link, which is the same way you are already accessing variable values in the `if` statements in the `steps` of the action you're proposing here.

please let me know if either of those explanations is unclear! 🙂

**Comment by GarvitSinghal47** [comment]
@thanksameeelian done the changes suggested by you.

**Comment by GarvitSinghal47** [comment]
@thanksameeelian can you help me why it is still failing , done the changes suggested by you even after that it is failing.

**Comment by thanksameeelian** [comment]
> @thanksameeelian can you help me why it is still failing , done the changes suggested by you even after that it is failing.

@GarvitSinghal47, is [any of the guidance provided in this stack overflow question](https://stackoverflow.com/a/6550543) helpful? 

in combination with this, i'm wondering if potentially the inclusion of the `-e` flag when Github Actions runs the script might [interact negatively with the`grep`ping you're trying to do in this action](https://stackoverflow.com/questions/9952177/whats-the-meaning-of-the-parameter-e-for-bash-shell-command-line)

**Comment by GarvitSinghal47** [comment]
> > @thanksameeelian can you help me why it is still failing , done the changes suggested by you even after that it is failing.
> 
> @GarvitSinghal47, is [any of the guidance provided in this stack overflow question](https://stackoverflow.com/a/6550543) helpful?
> 
> in combination with this, i'm wondering if potentially the inclusion of the `-e` flag when Github Actions runs the script might [interact negatively with the`grep`ping you're trying to do in this action](https://stackoverflow.com/questions/9952177/whats-the-meaning-of-the-parameter-e-for-bash-shell-command-line)

i have tried the || true method in the grep line know it is working fine and i have tested it locally as well when user exist and when not  but know the issue that is coming is coming in the commit and push method , can you tell me also what is wrong in it , below i have attached the act log and the updated code

updated code:
```
name: Add Contributor to AUTHORS.md
on:
  pull_request:
    types:
      - opened
      - synchronize

jobs:
  add_contributor:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v3

      - name: Check if contributor already exists
        id: check_contributor
        run: |
          exists=$(grep -c -F "$GITHUB_ACTOR" AUTHORS.md || true)
          if [[ $exists -gt 0 ]]; then
            echo "exists=true" >> $GITHUB_OUTPUT
          else
            echo "exists=false" >> $GITHUB_OUTPUT
          fi

      - name: Get contributor's full name
        id: get_full_name
        if: steps.check_contributor.outputs.exists == 'false'
        run: |
          username=$GITHUB_ACTOR
          access_token=$GITHUB_TOKEN
          full_name=$(curl -s -H "Authorization: token $access_token" "https://api.github.com/users/$username" | jq -r '.name')
          if [[ -z "$full_name" ]]; then
            full_name=$username
          fi
          echo "full_name=$full_name" >> $GITHUB_OUTPUT

      - name: Append contributor to AUTHORS.md
        id: append_contributor
        if: steps.check_contributor.outputs.exists == 'false'
        run: |
          full_name=${{ steps.get_full_name.outputs.full_name }}
          echo "| $full_name | $GITHUB_ACTOR |" >> AUTHORS.md

      - name: Commit and push changes
        if: steps.check_contributor.outputs.exists == 'false'
        run: |
          git config  user.email "action@github.com"
          git config  user.name "GitHub Action"
          git add AUTHORS.md
          git commit -m "Add $GITHUB_ACTOR to AUTHORS.md"
          git push
```

act log:
```
garvit@DESKTOP-E7IL2C3:/mnt/e/open-source/kolibri$ act -j add_contributor --bind --reuse
[Add Contributor to AUTHORS.md/add_contributor] 🚀  Start image=catthehacker/ubuntu:act-latest
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=true
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker create image=catthehacker/ubuntu:act-latest platform= entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker run image=catthehacker/ubuntu:act-latest platform= entrypoint=["tail" "-f" "/dev/null"] cmd=[]
[Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Checkout repository
[Add Contributor to AUTHORS.md/add_contributor]   ✅  Success - Main Checkout repository
[Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Check if contributor already exists
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/check_contributor] user= workdir=
[Add Contributor to AUTHORS.md/add_contributor]   ✅  Success - Main Check if contributor already exists
[Add Contributor to AUTHORS.md/add_contributor]   ⚙  ::set-output:: exists=false
[Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Get contributor's full name
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/get_full_name] user= workdir=
[Add Contributor to AUTHORS.md/add_contributor]   ✅  Success - Main Get contributor's full name
[Add Contributor to AUTHORS.md/add_contributor]   ⚙  ::set-output:: full_name=null
[Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Append contributor to AUTHORS.md
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/append_contributor] user= workdir=
[Add Contributor to AUTHORS.md/add_contributor]   ✅  Success - Main Append contributor to AUTHORS.md
[Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Commit and push changes
[Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/4] user= workdir=
| fatal: detected dubious ownership in repository at '/mnt/e/open-source/kolibri'
| To add an exception for this directory, call:
|
|       git config --global --add safe.directory /mnt/e/open-source/kolibri
[Add Contributor to AUTHORS.md/add_contributor]   ❌  Failure - Main Commit and push changes
[Add Contributor to AUTHORS.md/add_contributor] exitcode '128': failure
[Add Contributor to AUTHORS.md/add_contributor] 🏁  Job failed
Error: Job 'add_contributor' failed
```

and below is the line that is getting in authors,md after running above with using act
```
| null | nektos/act |
```

**Comment by MisRob** [comment]
Thanks @GarvitSinghal47 and @thanksameeelian for all your work on this, it's a valuable contribution.

@rtibbles mentioned that in some cases actions may be failing when their implementation is coming from a fork due to authentication. Looking at `| fatal: detected dubious ownership in repository at '/mnt/e/open-source/kolibri'` error in the log I wonder if this is the case here. @rtibbles Do we want to try to merge it and see if the action starts working after?

**Comment by GarvitSinghal47** [comment]
Thanks , one more thing before you check it after merge there is some update need to be done in code , I think I should just do the changes  for it then it would be good to go to check .

**Comment by MisRob** [comment]
Ah now it's failing for a new reason (which doesn't seem to be related to authorization)

```
Run git config  --local user.email "action@github.com"
[detached HEAD e3b2d6f] Add GarvitSinghal47 to AUTHORS.md
 1 file changed, 1 insertion(+)
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin HEAD:<name-of-remote-branch>
```

Since this action is running while the PR is open, I think we'd want to push to the branch from which a PR is open.

The good news is that this is the last step, so perhaps it will work soon :)!

**Comment by rtibbles** [comment]
This action is meant to be able to handle the committing for you, in another issue we have had an issue with it, because we are trying to trigger it from a non pull request event, but as we are using a pull request event here, it should work:

https://pre-commit.ci/lite.html

I think it should just commit any changes to the current PR branch.

**Comment by thanksameeelian** [comment]
i hope @MisRob & @rtibbles's comments are helpful, @GarvitSinghal47 :) 

once those problems are resolved, hopefully helped along by that feedback, i was wondering if - here or **in a follow-up issue** - we could adjust the script to catch & address instances of `null` getting added to the `AUTHORS.md` table. i noted that in the logs you shared (a selection of which i've re-shared below), `null` was added as the author's name that made it into the `AUTHORS.md` file, and that `null` value was captured & logged early in the process of running the action. even though the actual failure that caused the script to exit occurred after this, this is another problem we're not currently handling in an ideal way.

i'm curious whether we'd prefer to have the table have a specific placeholder if the contributor's name information isn't available but their username is (e.g. "N/A", "Not available" for full name instead), whether keeping it to `null` is fine, or whether we'd like to skip adding that person entirely. interested to hear your thoughts on this, @GarvitSinghal47, @rtibbles & @MisRob, and again this doesn't have to be a _now_ thing, but i'd love to create a secondary issue for it (or track it in another way) if it won't be addressed here. 

> act log:
> 
> ```
> ...
>
> [Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Get contributor's full name
> [Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/get_full_name] user= workdir=
> [Add Contributor to AUTHORS.md/add_contributor]   ✅  Success - Main Get contributor's full name
> [Add Contributor to AUTHORS.md/add_contributor]   ⚙  ::set-output:: full_name=null
>
> ...
>
> [Add Contributor to AUTHORS.md/add_contributor] ⭐ Run Main Commit and push changes
> [Add Contributor to AUTHORS.md/add_contributor]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /var/run/act/workflow/4] user= workdir=
> | fatal: detected dubious ownership in repository at '/mnt/e/open-source/kolibri'
> | To add an exception for this directory, call:
> |
> |       git config --global --add safe.directory /mnt/e/open-source/kolibri
> [Add Contributor to AUTHORS.md/add_contributor]   ❌  Failure - Main Commit and push changes
> [Add Contributor to AUTHORS.md/add_contributor] exitcode '128': failure
> [Add Contributor to AUTHORS.md/add_contributor] 🏁  Job failed
> Error: Job 'add_contributor' failed
> ```
> 
> and below is the line that is getting in authors,md after running above with using act
> 
> ```
> | null | nektos/act |
> ```

**Comment by rtibbles** [comment]
@thanksameeelian definitely not good to have `null` in the AUTHORS.md - I think the intended (and desirable) behaviour in this PR is to instead reuse the github username if the full name is not available. It seems that this is not happening at the moment. Maybe needs a check against `null` rather than just assuming it will be a falsey value?

**Comment by MisRob** [comment]
> I think the intended (and desirable) behaviour in this PR is to instead reuse the github username if the full name is not available

@rtibbles Any reason why not to simply leave "Name" entry empty or use "-" when we don't have the full name so it corresponds to the situation? When I try to imagine reusing the username, it ends up looking a bit strange and may even give the impression of an error in the script, rather than intended behavior.


| Name | Github user |
|--|-------------|
| 66eli77 | 66eli77 |
| Adam Stasiw | AdamStasiw |

In comparison to

| Name | Github user |
|--|-------------|
| - | 66eli77 |
| Adam Stasiw | AdamStasiw |

or 

| Name | Github user |
|--|-------------|
|  | 66eli77 |
| Adam Stasiw | AdamStasiw |

In any case, I agree it'd be nice not to call people `null` :)

**Comment by rtibbles** [comment]
Mostly just to be consistent with how it already is - but it's not hugely important - not `null` is the important thing!

**Comment by MisRob** [comment]
@rtibbles Okay, thank you. I'd then suggest using "-" as an empty name placeholder for now. Then we could do a cleanup of old entries to address your concern about consistency at some point (not in this PR). I will put it on my todos list around the open-source community. Does that sound good?

**Comment by rtibbles** [comment]
Yep!

**Comment by GarvitSinghal47** [comment]
so there currently 3 issue that need to be resolved.
1. Replace null name with - .
2. add contributor to authors.md only when the pr is merged .
3. I think the pr not need to be pushed as well just need to be commit it to the branch it was made.

Just want to confirm these all are the issue from others that need to be resolved or i missed something.

Will do the changes according in the code and will create a commit for it soon.

**Comment by MisRob** [comment]
Thank you for your patience and continuous cooperation @GarvitSinghal47

**Comment by MisRob** [comment]
@GarvitSinghal47 

> so there currently 3 issue that need to be resolved.

When it comes to my feedback, I can confirm "1. Replace null name with - "

Will defer to @rtibbles when it comes to 2. and 3.

**Comment by GarvitSinghal47** [comment]
> @GarvitSinghal47 
> 
> > so there currently 3 issue that need to be resolved.
> 
> When it comes to my feedback, I can confirm "1. Replace null name with - "
> 
> Will defer to @rtibbles when it comes to 2. and 3.


Okk

**Comment by rtibbles** [comment]
I am not completely sure where the motivation for:

> 2. add contributor to authors.md only when the pr is merged .

is coming from - I think it is fine as is to add it to the currently open PR branch (which also makes it easier to test in the context of this Pull Request). The issue with the current implementation as I see it, is that the commit with the update to AUTHORS.md would get added to the PR branch after merge, so it will never make it into the target branch!

For:

> 3. I think the pr not need to be pushed as well just need to be commit it to the branch it was made.

I think the issue here is with how to best push from the checkout that has happened already and then resolving that into a push to the branch. I suspect instead of trying to handle turning the detached HEAD checkout into a branch that we can commit to, using this action might save some work: https://github.com/pre-commit-ci/lite-action/tree/main might be better, as it seems to have handled this complexity already.

**Comment by GarvitSinghal47** [comment]
> 

The motivation behind it initially was to make it run not from the main branch instead make it to run from the individual  branch after the pr is merged.

@rtibbles initially i was thinking to make it happen to work when the pr is merged but as mentioned above it will make difficult to test it and also  it will never make it into the target branch! , so i will change it back to the previous one code and also for 3 part will implement using the method mentioned above .

**Comment by GarvitSinghal47** [comment]
@rtibbles done the changes

**Comment by GarvitSinghal47** [comment]
@rtibbles there is something still not working as intended as no commit is happening as the precommit  is not pushing in into the current branch and i also think the use of precommit is to just check out the linting error , i think need to do something else to handle the push as till commit code is running fine .

i find something online for it just want to confirm that will it work or not .

```
- name: Create PR
        id: create_pr
        if: steps.check_contributor.outputs.exists == 'false'
        uses: peter-evans/create-pull-request@v3
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
          title: "Add $GITHUB_ACTOR to AUTHORS.md"
          body: "This pull request adds $GITHUB_ACTOR to the AUTHORS.md file."
          branch: ${{ github.event.pull_request.head.ref }}
         
 ```

**Comment by GarvitSinghal47** [comment]
@rtibbles i have tried every way possible that i find online to push after setting up the worflow in a repo of mine , but i am unable to find a way to push it to the current user branch as a commit using github action.

The only way i am finding is that we need to setup a PAT token or SSH key from the user side first for the repo , then after that it is possible only to trigger a commit from the current workflow in the repo.

As majority of push operation that are present online was to create a new pull request in the main repo from which the workflow has been runned in or using token.

And i have also find a article to achieve the same thing that we want to do and i have tried it also but it is also not working.

Article Link- https://github.com/orgs/community/discussions/24945

@thanksameeelian can you help me to figure out how i can achieve it or any assistance you can provide me in it.

**Comment by rtibbles** [comment]
Hi @GarvitSinghal47 - thanks for all the work here - the good news is that you seem to have resolved all the previous technical issues with reading and setting the username and full name, so that's great!

It seems like the issue remaining may have more to do with authentication issues than anything else, so this may be the point where we have to step in as the repo owners to resolve things. I'll check the PR out and try some tweaks to see if I can fix the authentication issues here - this will involve me pushing to your PR, so just a heads up that will happen!

**Comment by rtibbles** [comment]
Hi @GarvitSinghal47 - I have made an update that I think might have fixed your earlier issue - unfortunately, it didn't run because I am already in the AUTHORS.md. Could you pull the changes and merge in latest develop so that it gets a push event with your username attached?

**Comment by GarvitSinghal47** [comment]
> Hi @GarvitSinghal47 - I have made an update that I think might have fixed your earlier issue - unfortunately, it didn't run because I am already in the AUTHORS.md. Could you pull the changes and merge in latest develop so that it gets a push event with your username attached?

Sure

**Comment by rtibbles** [comment]
Unfortunately, I think we are now hitting a permissions issue which is preventing the commit from being pushed. I'm going to merge this as is, and then debug further issues there - thanks for all the hard work here @GarvitSinghal47!

**Review by rtibbles — APPROVED** [review]

**Comment by thanksameeelian** [comment]
i just want to chime in to also thank you so much for all your work on this, @GarvitSinghal47!! you reacted to all our feedback with great flexibility & quick turn-around times and we really appreciate your contribution! 🙂

**Comment by GarvitSinghal47** [comment]
> i just want to chime in to also thank you so much for all your work on this, @GarvitSinghal47!! you reacted to all our feedback with great flexibility & quick turn-around times and we really appreciate your contribution! 🙂

Thanks for your kind word.

**Pull request merged**